File: packages/flow_widget_android/android/src/main/kotlin/dev/flowwidget/android/FlowWidgetValueCodec.kt
package dev.flowwidget.android

import android.util.Base64
import org.json.JSONArray
import org.json.JSONObject

/**
 * Encodes and decodes [FlowWidgetValue] wire maps `{t, v}` for MethodChannel
 * transport and JSON persistence.
 */
object FlowWidgetValueCodec {
    private const val KEY_TYPE = "t"
    private const val KEY_VALUE = "v"

    fun encode(wire: Map<*, *>): Map<String, Any?> {
        val type = wire[KEY_TYPE] as? String
            ?: throw IllegalArgumentException("Missing wire type discriminator")
        val value = wire[KEY_VALUE]
        return mapOf(KEY_TYPE to type, KEY_VALUE to encodeValue(type, value))
    }

    fun decode(wire: Map<*, *>): Map<String, Any?> {
        val type = wire[KEY_TYPE] as? String
            ?: throw IllegalArgumentException("Missing wire type discriminator")
        val value = wire[KEY_VALUE]
        return mapOf(KEY_TYPE to type, KEY_VALUE to decodeValue(type, value))
    }

    fun encodeToJson(wire: Map<*, *>): String {
        return JSONObject(encode(wire)).toString()
    }

    fun decodeFromJson(json: String): Map<String, Any?> {
        val objectJson = JSONObject(json)
        val wire = mapOf(
            KEY_TYPE to objectJson.getString(KEY_TYPE),
            KEY_VALUE to objectJson.opt(KEY_VALUE),
        )
        return decode(wire)
    }

    fun encodeDataMap(data: Map<*, *>?): Map<String, Map<String, Any?>> {
        if (data == null) return emptyMap()
        return buildMap {
            for ((key, value) in data) {
                val stringKey = key as? String ?: continue
                val wire = value as? Map<*, *> ?: continue
                put(stringKey, encode(wire))
            }
        }
    }

    fun decodeDataMap(data: Map<*, *>?): Map<String, Map<String, Any?>> {
        if (data == null) return emptyMap()
        return buildMap {
            for ((key, value) in data) {
                val stringKey = key as? String ?: continue
                val wire = value as? Map<*, *> ?: continue
                put(stringKey, decode(wire))
            }
        }
    }

    private fun encodeValue(type: String, value: Any?): Any? {
        return when (type) {
            "m" -> encodeMap(value)
            "l" -> encodeList(value)
            "bin" -> encodeBytes(value)
            else -> value
        }
    }

    private fun decodeValue(type: String, value: Any?): Any? {
        return when (type) {
            "m" -> decodeMap(value)
            "l" -> decodeList(value)
            "bin" -> decodeBytes(value)
            else -> value
        }
    }

    private fun encodeMap(value: Any?): JSONObject {
        val map = value as? Map<*, *> ?: emptyMap<Any?, Any?>()
        val json = JSONObject()
        for ((key, entryValue) in map) {
            val stringKey = key as? String ?: continue
            val wire = entryValue as? Map<*, *> ?: continue
            json.put(stringKey, JSONObject(encode(wire)))
        }
        return json
    }

    private fun decodeMap(value: Any?): Map<String, Any?> {
        val json = value as? JSONObject ?: return emptyMap()
        return buildMap {
            val keys = json.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                val nested = json.getJSONObject(key)
                put(key, decodeValue(nested.getString(KEY_TYPE), nested.opt(KEY_VALUE)))
            }
        }
    }

    private fun encodeList(value: Any?): JSONArray {
        val list = value as? List<*> ?: emptyList<Any?>()
        val json = JSONArray()
        for (item in list) {
            val wire = item as? Map<*, *> ?: continue
            json.put(JSONObject(encode(wire)))
        }
        return json
    }

    private fun decodeList(value: Any?): List<Any?> {
        val json = value as? JSONArray ?: return emptyList()
        return buildList {
            for (index in 0 until json.length()) {
                val nested = json.getJSONObject(index)
                add(decodeValue(nested.getString(KEY_TYPE), nested.opt(KEY_VALUE)))
            }
        }
    }

    private fun encodeBytes(value: Any?): String {
        return when (value) {
            is ByteArray -> Base64.encodeToString(value, Base64.NO_WRAP)
            is List<*> -> {
                val bytes = ByteArray(value.size) { index ->
                    (value[index] as Number).toInt().toByte()
                }
                Base64.encodeToString(bytes, Base64.NO_WRAP)
            }
            is String -> value
            else -> ""
        }
    }

    private fun decodeBytes(value: Any?): ByteArray {
        val encoded = value as? String ?: return ByteArray(0)
        return Base64.decode(encoded, Base64.NO_WRAP)
    }
}

File: packages/flow_widget_macos/macos/flow_widget_macos/Sources/flow_widget_macos/FlowWidgetValueCodec.swift
import Foundation
import FlutterMacOS

/// Encodes and decodes flow_widget wire values `{t, v}`.
enum FlowWidgetValueCodec {
  static func encode(_ value: Any) -> [String: Any] {
    switch value {
    case let string as String:
      return ["t": "s", "v": string]
    case let int as Int:
      return ["t": "i", "v": int]
    case let int64 as Int64:
      return ["t": "i", "v": Int(int64)]
    case let double as Double:
      return ["t": "d", "v": double]
    case let float as Float:
      return ["t": "d", "v": Double(float)]
    case let bool as Bool:
      return ["t": "b", "v": bool]
    case let date as Date:
      return ["t": "dt", "v": Int(date.timeIntervalSince1970 * 1000)]
    case let data as Data:
      return ["t": "bin", "v": FlutterStandardTypedData(bytes: data)]
    case let map as [String: Any]:
      var encoded: [String: Any] = [:]
      for (key, nested) in map {
        encoded[key] = encode(nested)
      }
      return ["t": "m", "v": encoded]
    case let list as [Any]:
      return ["t": "l", "v": list.map { encode($0) }]
    case let wire as [String: Any] where wire["t"] != nil:
      return wire
    default:
      return ["t": "s", "v": String(describing: value)]
    }
  }

  static func decode(_ wire: [String: Any]) -> Any? {
    guard let type = wire["t"] as? String else { return nil }
    let value = wire["v"]
    switch type {
    case "s":
      return value as? String
    case "i":
      if let int = value as? Int { return int }
      if let int64 = value as? Int64 { return Int(int64) }
      if let num = value as? NSNumber { return num.intValue }
      return nil
    case "d":
      if let double = value as? Double { return double }
      if let num = value as? NSNumber { return num.doubleValue }
      return nil
    case "b":
      return value as? Bool
    case "dt":
      let millis: Int
      if let int = value as? Int {
        millis = int
      } else if let int64 = value as? Int64 {
        millis = Int(int64)
      } else if let num = value as? NSNumber {
        millis = num.intValue
      } else {
        return nil
      }
      return Date(timeIntervalSince1970: TimeInterval(millis) / 1000.0)
    case "j":
      return value as? String
    case "bin":
      if let typed = value as? FlutterStandardTypedData {
        return typed.data
      }
      if let data = value as? Data {
        return data
      }
      return nil
    case "m":
      guard let raw = value as? [String: Any] else { return nil }
      var decoded: [String: Any] = [:]
      for (key, nested) in raw {
        if let nestedWire = nested as? [String: Any], let decodedValue = decode(nestedWire) {
          decoded[key] = decodedValue
        }
      }
      return decoded
    case "l":
      guard let raw = value as? [Any] else { return nil }
      return raw.compactMap { item -> Any? in
        guard let itemWire = item as? [String: Any] else { return nil }
        return decode(itemWire)
      }
    default:
      return nil
    }
  }

  static func decodeMap(_ wire: [String: Any]) -> [String: Any] {
    var result: [String: Any] = [:]
    for (key, nested) in wire {
      if let nestedWire = nested as? [String: Any], let decoded = decode(nestedWire) {
        result[key] = decoded
      }
    }
    return result
  }

  static func encodeMap(_ map: [String: Any]) -> [String: Any] {
    var result: [String: Any] = [:]
    for (key, value) in map {
      result[key] = encode(value)
    }
    return result
  }
}
